README.md
# Minimalist µServices

[![NPM Version][npm-badge]][npm-url]
[![Dependencies][david-badge]][david-url]
[![Build][travis-badge]][travis-url]
[![Test Coverage][coveralls-badge]][coveralls-url]
[![License: MIT][license-badge]][license-url]

The goal of [mini-service][mini-service-url] is to give the minimal structure to implement a µService, that can be invoked locally or remotely.
[Mini-client][mini-client-url] is a generic client that [limits coupling][distributed-monolith] between your project and a remote mini-service.

Its principles are the following:
- very easy to add new service api endpoints
- easy to use client interface, same usage both locally and remotely
- hide deployment details and provide simple-yet-working solution
- promises based

All mini-* libraries use the latest ES2017 features, so they requires node 8+

Please checkout the [API reference][api-reference]


This project was kindly sponsored by [nearForm][nearform].

## License

Copyright [Damien Simonin Feugas][feugy] and other contributors, licensed under [MIT](./LICENSE).

[nearform]: http://nearform.com
[feugy]: https://github.com/feugy
[mini-service-url]: https://github.com/feugy/mini-service
[mini-client-url]: https://github.com/feugy/mini-client
[david-badge]: https://img.shields.io/david/feugy/mini-utils.svg
[david-url]: https://david-dm.org/feugy/mini-utils
[npm-badge]: https://img.shields.io/npm/v/mini-service-utils.svg
[npm-url]: https://npmjs.org/package/mini-service-utils
[travis-badge]: https://api.travis-ci.org/feugy/mini-utils.svg
[travis-url]: https://travis-ci.org/feugy/mini-utils
[coveralls-badge]: https://img.shields.io/coveralls/feugy/mini-utils/master.svg
[coveralls-url]: https://coveralls.io/r/feugy/mini-utils?branch=master
[nsp-badge]: https://nodesecurity.io/orgs/perso/projects/6bc9b474-6f9e-4db0-a4d3-c3bf5443a63a/badge
[nsp-url]: https://nodesecurity.io/orgs/perso/projects/6bc9b474-6f9e-4db0-a4d3-c3bf5443a63a
[distributed-monolith]: https://www.infoq.com/news/2016/02/services-distributed-monolith
[api-reference]: https://feugy.github.io/mini-utils/
[license-badge]: https://img.shields.io/badge/License-MIT-green.svg
[license-url]: https://github.com/feugy/mini-utils/blob/master/LICENSE


CHANGELOG.md
# Changelog

## 4.0.1
#### Changed
- Use latest standard formatting
- Remove NSP since it was decomissioned
- Use yarn instead of NPM
- Dependencies update

## 4.0.0
#### Added
- `loadTransport()` which validates incoming options and loads selected transport 
- `optionSchema` Joi schema used to validate incoming options

#### Changed
- Reformat CHANGELOG to follow [Keep a Changelog](https://keepachangelog.com) recommandations
- New documentation with latest docma v2.0.0
- Dependencies update

## 3.0.1
#### Fixed
- `getParamNames()` to support destructured parameters with default values

## 3.0.0
#### Fixed
- `getParamNames()` supports async functions with single parameter shortcut: `
   ```js
   getParamNames(async a => {}) === ['a']
   ```

#### Changed
- **Breaking**: `getParamNames()` supports rest parameters:
   ```js
   getParamNames((a, b, ...rest) => {}) === ['a', 'b', 'rest']
   ```
   It used to throw an Error
- **Breaking**: `getParamNames()` supports destructured parameters:
   ```js
   getParamNames((a, {b: [c]}, [d, e], f) => {}) === ['a', 'param1', 'param2', 'f']
   ```
   It used to throw an Error
- **Breaking**: Removed: `validateParams()` deprecated function


## 2.4.0
#### Changed
- **Deprecation**: `validateParams()` will be deprecated in next major version, in favour of: `enrichError()`
   Validation of input parameters and returned result is now responsability of mini-client & mini-service.
   `enrichError()` ensures that Joi `ValidationError`, when thrown, get enriched and turned into consistent
   Boom errors
- Dependencies update


## 2.3.0
### Added
- Expose CRC32 checksum header name

#### Changed
- Dependencies update


## 2.2.1
### Added
- `extractValidate()` function allowing validation objects in options
- Include API name in validation errors

#### Changed
- Better documentation


## 2.0.0
### Added
- Introduce `extractGroup()`

#### Changed
- **Breaking**: Rename service to group
- Use husky instead of ghooks
- Dependencies update


## 1.0.0
### Added
- initial release

test/utils.js
const Lab = require('lab')
const Joi = require('joi')
const { badRequest } = require('boom')
const assert = require('power-assert')
const rewire = require('rewire')
const utils = rewire('../')

const lab = exports.lab = Lab.script()
const { describe, it, beforeEach } = lab

describe('Utilities', () => {
  /* eslint no-empty-function: 0 */
  const noop = () => {}

  describe('extractValidate', () => {
    it('should extract from api.validate', () => {
      const id = 'test'
      const apis = {
        [id]: () => {}
      }
      const validation = Joi.array()
      apis[id].validate = validation
      assert.deepStrictEqual(utils.extractValidate(id, apis), validation)
    })

    it('should extract another clause from api', () => {
      const id = 'test5'
      const apis = {
        [id]: () => {}
      }
      const validation = Joi.array()
      apis[id].responseSchema = validation
      assert.deepStrictEqual(utils.extractValidate(id, apis, {}, 'responseSchema'), validation)
    })

    it('should extract from opts.validate', () => {
      const id = 'test2'
      const apis = {
        [id]: () => {}
      }
      const validation = Joi.array()
      assert.deepStrictEqual(utils.extractValidate(id, apis, { validate: validation }), validation)
    })

    it('should extract another clause from opts', () => {
      const id = 'test6'
      const apis = {
        [id]: () => {}
      }
      const validation = Joi.array()
      assert.deepStrictEqual(
        utils.extractValidate(id, apis, { responseSchema: validation }, 'responseSchema'),
        validation
      )
    })

    it('should not fail on missing validate key', () => {
      const id = 'test3'
      const apis = {
        [id]: () => {}
      }
      assert.strictEqual(utils.extractValidate(id, apis, {}), null)
    })

    it('should not fail on missing options', () => {
      const id = 'test4'
      const apis = {
        [id]: () => {}
      }
      assert.strictEqual(utils.extractValidate(id, apis), null)
    })
  })

  describe('getLogger', () => {
    beforeEach(() => {
      utils.__set__('logger', null)
    })

    it('should create a bunyan logger', () => {
      const logger = utils.getLogger()
      assert(logger)
      assert(logger.debug)
      assert.equal(logger.fields.name, 'mini-service-utils')
    })

    it('should reuse the same instance', () => {
      assert.strictEqual(utils.getLogger(), utils.getLogger())
    })

    it('should be customizable', () => {
      const logger = utils.getLogger({ name: 'toto' })
      assert(logger)
      assert(logger.debug)
      assert.equal(logger.fields.name, 'toto')
    })
  })

  describe('arrayToObj', () => {
    it('should transform an array to an object', () => {
      assert.deepStrictEqual(utils.arrayToObj([1, 2], ['count', 'other']), { count: 1, other: 2 })
    })

    it('should keep unexpected items in result', () => {
      assert.deepStrictEqual(utils.arrayToObj([1, 2, 3, 'haha'], ['count', 'other']), {
        count: 1,
        other: 2,
        2: 3,
        3: 'haha'
      })
    })
  })

  describe('isApi', () => {
    it('should allow empty object', () => {
      assert(utils.isApi({}))
    })

    it('should allow plain object', () => {
      assert(utils.isApi({ method: () => {}, attribute: 1 }))
    })

    it('should allow class instances', () => {
      class Service {
        method () {}
      }
      assert(utils.isApi(new Service()))
    })

    it('should not allow falsy values', () => {
      assert.equal(utils.isApi(), false)
      assert.equal(utils.isApi(null), false)
      assert.equal(utils.isApi(''), false)
      assert.equal(utils.isApi(false), false)
      assert.equal(utils.isApi(0), false)
    })

    it('should not allow string and numbers values', () => {
      assert.equal(utils.isApi('hi'), false)
      assert.equal(utils.isApi(10), false)
    })

    it('should not allow arrays', () => {
      assert.equal(utils.isApi([]), false)
      assert.equal(utils.isApi([1, 2]), false)
    })
  })

  describe('enrichError', () => {
    const validationError = Joi.object({ other: Joi.string().required() }).validate({ count: 1 }).error

    it('should return null if no error', () => {
      assert(utils.enrichError(null, '') === null)
    })

    it('should return bad request on input validation error', () => {
      const error = utils.enrichError(validationError, 'test-input')
      assert(error.isBoom)
      assert(error.output.payload.statusCode === 400)
      assert(error.output.payload.error === 'Bad Request')
      assert(error.message.includes('Incorrect parameters for API test-input'))
      assert(error.message.includes('"other" is required]'))
    })

    it('should return bad response on result validation error', () => {
      const error = utils.enrichError(validationError, 'test-response', false)
      assert(error.isBoom)
      assert(error.output.payload.statusCode === 512)
      assert(error.output.payload.error === 'Bad Response')
      assert(error.message.includes('Incorrect response for API test-response'))
      assert(error.message.includes('"other" is required]'))
    })

    it('should override boom error', () => {
      const error = utils.enrichError(badRequest(validationError), 'test-override', false)
      assert(error.isBoom)
      assert(error.output.payload.statusCode === 512)
      assert(error.output.payload.error === 'Bad Response')
      assert(error.message.includes('Incorrect response for API test-override'))
      assert(error.message.includes('"other" is required]'))
    })
  })

  describe('getParamNames', () => {
    it('should fails on null', () => {
      assert.throws(() => utils.getParamNames(null), /unsupported function null/)
    })

    it('should fails on undefined', () => {
      assert.throws(() => utils.getParamNames(), /unsupported function undefined/)
    })

    it('should fails on object', () => {
      assert.throws(() => utils.getParamNames({}), /unsupported function \[object Object\]/)
    })

    const suites = [{
      name: 'without parameter',
      expected: [],
      tests: [{
        name: 'function declaration',
        code: () => {
          function declared () {}
          return declared
        }
      }, {
        name: 'async function declaration',
        code: () => {
          async function declared () {}
          return declared
        }
      }, {
        name: 'anonymous function',
        code: () => function () {}
      }, {
        name: 'async anonymous function',
        code: () => async function () {}
      }, {
        name: 'named function',
        code: () => function named () {}
      }, {
        name: 'async named function',
        code: () => async function named () {}
      }, {
        name: 'function shortcut',
        code: () => {
          const obj = {
            named () {}
          }
          return obj.named
        }
      }, {
        name: 'async function shortcut',
        code: () => {
          const obj = {
            async named () {}
          }
          return obj.named
        }
      }, {
        name: 'arrow function',
        code: () => () => {}
      }, {
        name: 'async arrow function',
        code: () => async () => {}
      }]
    }, {
      name: 'with single parameter',
      expected: ['a'],
      tests: [{
        name: 'function declaration',
        code: () => {
          function declared (a) {}
          return declared
        }
      }, {
        name: 'async function declaration',
        code: () => {
          async function declared (a) {}
          return declared
        }
      }, {
        name: 'anonymous function',
        code: () => function (a) {}
      }, {
        name: 'async anonymous function',
        code: () => async function (a) {}
      }, {
        name: 'named function',
        code: () => function named (a) {}
      }, {
        name: 'async named function',
        code: () => async function named (a) {}
      }, {
        name: 'function shortcut',
        code: () => {
          const obj = {
            named (a) {}
          }
          return obj.named
        }
      }, {
        name: 'async function shortcut',
        code: () => {
          const obj = {
            async named (a) {}
          }
          return obj.named
        }
      }, {
        name: 'arrow function',
        code: () => a => {}
      }, {
        name: 'async arrow function',
        code: () => async a => {}
      }]
    }, {
      name: 'with multiple parameters',
      expected: ['a', 'b', 'c'],
      tests: [{
        name: 'function declaration',
        code: () => {
          function declared (a, b, c) {}
          return declared
        }
      }, {
        name: 'async function declaration',
        code: () => {
          async function declared (a, b, c) {}
          return declared
        }
      }, {
        name: 'anonymous function',
        code: () => function (a, b, c) {}
      }, {
        name: 'async anonymous function',
        code: () => async function (a, b, c) {}
      }, {
        name: 'named function',
        code: () => function named (a, b, c) {}
      }, {
        name: 'async named function',
        code: () => async function named (a, b, c) {}
      }, {
        name: 'function shortcut',
        code: () => {
          const obj = {
            named (a, b, c) {}
          }
          return obj.named
        }
      }, {
        name: 'async function shortcut',
        code: () => {
          const obj = {
            async named (a, b, c) {}
          }
          return obj.named
        }
      }, {
        name: 'arrow function',
        code: () => (a, b, c) => {}
      }, {
        name: 'async arrow function',
        code: () => async (a, b, c) => {}
      }]
    }, {
      name: 'with default values',
      expected: ['a', 'b', 'c'],
      tests: [{
        name: 'function declaration',
        code: () => {
          function declared (a, b, c = null) {}
          return declared
        }
      }, {
        name: 'async function declaration',
        code: () => {
          async function declared (a, b, c = 10) {}
          return declared
        }
      }, {
        name: 'anonymous function',
        code: () => function (a, b, c = '') {}
      }, {
        name: 'async anonymous function',
        code: () => async function (a, b, c = []) {}
      }, {
        name: 'named function',
        code: () => function named (a, b, c = true) {}
      }, {
        name: 'async named function',
        code: () => async function named (a, b, c = false) {}
      }, {
        name: 'function shortcut',
        code: () => {
          const obj = {
            named (a, b, c = {}) {}
          }
          return obj.named
        }
      }, {
        name: 'async function shortcut',
        code: () => {
          const obj = {
            async named (a, b, c = () => {}) {}
          }
          return obj.named
        }
      }, {
        name: 'arrow function',
        code: () => (a, b, c = describe) => {}
      }, {
        name: 'async arrow function',
        code: () => async (a, b, c = null) => {}
      }]
    }, {
      name: 'with rest parameters',
      expected: ['a', 'b', 'rest'],
      tests: [{
        name: 'function declaration',
        code: () => {
          function declared (a, b, ...rest) {}
          return declared
        }
      }, {
        name: 'async function declaration',
        code: () => {
          async function declared (a, b, ...rest) {}
          return declared
        }
      }, {
        name: 'anonymous function',
        code: () => function (a, b, ...rest) {}
      }, {
        name: 'async anonymous function',
        code: () => async function (a, b, ...rest) {}
      }, {
        name: 'named function',
        code: () => function named (a, b, ...rest) {}
      }, {
        name: 'async named function',
        code: () => async function named (a, b, ...rest) {}
      }, {
        name: 'function shortcut',
        code: () => {
          const obj = {
            named (a, b, ...rest) {}
          }
          return obj.named
        }
      }, {
        name: 'async function shortcut',
        code: () => {
          const obj = {
            async named (a, b, ...rest) {}
          }
          return obj.named
        }
      }, {
        name: 'arrow function',
        code: () => (a, b, ...rest) => {}
      }, {
        name: 'async arrow function',
        code: () => async (a, b, ...rest) => {}
      }]
    }, {
      name: 'with destructured parameters',
      expected: ['a', 'param2', 'c'],
      tests: [{
        name: 'function declaration',
        code: () => {
          function declared (a, { b: [d] }, c) {}
          return declared
        }
      }, {
        name: 'async function declaration',
        code: () => {
          async function declared (a, { b: { d } }, c) {}
          return declared
        }
      }, {
        name: 'anonymous function',
        code: () => function (a, { b, d }, c) {}
      }, {
        name: 'async anonymous function',
        code: () => async function (a, [b], c) {}
      }, {
        name: 'named function',
        code: () => function named (a, { b }, c) {}
      }, {
        name: 'async named function',
        code: () => async function named (a, { b }, c) {}
      }, {
        name: 'function shortcut',
        code: () => {
          const obj = {
            named (a, [b, ...rest], c) {}
          }
          return obj.named
        }
      }, {
        name: 'async function shortcut',
        code: () => {
          const obj = {
            async named (a, { b } = {}, c) {}
          }
          return obj.named
        }
      }, {
        name: 'arrow function',
        code: () => (a, [b] = [], c) => {}
      }, {
        name: 'async arrow function',
        code: () => async (a, { b }, c) => {}
      }]
    }]

    suites.forEach(({ name, expected, tests }) => {
      describe(name, () => {
        tests.forEach(({ name, code }) => {
          it(`should handle ${name}`, () => {
            assert.deepStrictEqual(utils.getParamNames(code()), expected)
          })
        })
      })
    })
  })

  describe('extractGroups', () => {
    it('should get group from parameters', () => {
      const name = 'test'
      const init = noop
      assert.deepStrictEqual(utils.extractGroups({ name, init }), {
        groups: [{ name, init }],
        groupOpts: { [name]: { name, init } }
      })
    })

    it('should get groups from property', () => {
      const name1 = 'test1'
      const init1 = noop
      const name2 = 'test2'
      const init2 = noop
      assert.deepStrictEqual(utils.extractGroups({
        groups: [{ name: name1, init: init1 }, { name: name2, init: init2 }]
      }), {
        groups: [{ name: name1, init: init1 }, { name: name2, init: init2 }],
        groupOpts: {}
      })
    })

    it('should get group options from property', () => {
      const name1 = 'test1'
      const init1 = noop
      const name2 = 'test2'
      const init2 = noop
      assert.deepStrictEqual(utils.extractGroups({
        groups: [{ name: name1, init: init1 }, { name: name1, init: init1 }],
        groupOpts: { [name1]: { name: name1 }, [name2]: { name: name2 } }
      }), {
        groups: [{ name: name1, init: init1 }, { name: name1, init: init2 }],
        groupOpts: { [name1]: { name: name1 }, [name2]: { name: name2 } }
      })
    })

    it('should fail if no group found', () => {
      assert.throws(() => utils.extractGroups({ name: 'test' }), /No APIs nor APIs groups defined/)
    })

    it('should fail on invalid group found', () => {
      assert.throws(() => utils.extractGroups({ groups: ['not a group'] }), /Group definition at position 0/)
    })
  })

  describe('loadTransport', () => {
    const logger = {
      debug: () => {},
      info: () => {}
    }

    it('should assert incoming options', () => {
      assert.throws(() => {
        utils.loadTransport({ transport: { type: 'Whatever' } }, require)
      }, /"logger" is required/)
    })

    it('should report unknown transport', () => {
      assert.throws(() => {
        utils.loadTransport({ transport: { type: 'unknown' }, logger }, require)
      }, /Cannot load transport unknown/)
    })

    it('should report existing transport', () => {
      assert.strictEqual(utils.loadTransport({ transport: { type: 'test' }, logger }, require), 'yeah')
    })
  })
})


test/test-utils.js
const { getLogger } = require('../')

// store logger levels for further restoration
let loggerLevels = null

/**
 * Shutdown (or changes levels of) all logger outputs
 * @returns {Promise} promise - resolve when logger has been shot down
 */
exports.shutdownLogger = () =>
  new Promise(resolve => {
    loggerLevels = getLogger().levels()
    getLogger().level(Number.POSITIVE_INFINITY)
    resolve()
  })

/**
 * Restore levels of existing logger outputs
 * @returns {Promise} promise - resolve when logger levels have been restored
 */
exports.restoreLogger = () =>
  new Promise(resolve => {
    for (const { level, i } of loggerLevels.entries()) {
      getLogger().level(i, level)
    }
    resolve()
  })


test/transports/test.js
/* fixture for loadTransports() */
module.exports = 'yeah'


lib/index.js
/**
 * Utilities for mini-service and mini-client
 * @module mini-service-utils
 */
const bunyan = require('bunyan')
const joi = require('joi')
const { boomify } = require('boom')
const acorn = require('acorn-loose')
const { name } = require('../package.json')

/**
 * Joi schema used for validations.
 * @typedef {Object} Joi
 * @see {@link https://github.com/hapijs/joi/blob/v13.1.2/API.md|Joi API}
 */

/**
 * Bunyan compatible logger
 * @typedef {Object} Bunyan
 * @see {@link https://github.com/trentm/node-bunyan#log-method-api|Bunyan API}
 */

/**
 * List of extracted groups with their parameters
 * @typedef {Object} ExtractedGroups
 * @property {Array<Object>} groups   - array of API groups
 * @property {Object} groupOpts       - hash of group parameter, group names are used as properties
 */

// already created logger
let logger = null

/**
 * Creates or re-use an existing logger
 * @param {Object} [opts = {}]  - optional buyan options
 * @returns {Bunyan}            logger object.
 */
exports.getLogger = (opts = {}) => {
  if (!logger) {
    logger = bunyan.createLogger(Object.assign({
      name,
      streams: [{
        level: 'debug',
        stream: process.stdout
      }]
    }, opts))
  }
  return logger
}

/**
 * Utility function to get parameter name from AST node.
 * Supports:
 *
 * - plain identifier
 * - default values (assignment pattern & expression)
 * - rest parameters (rest & spread element)
 * - destructured parameters (object pattern & expression, array pattern & expression)
 *   who got a generated name
 *
 * @param {Object} node   - AST node, likely to represent a function parameter
 * @param {Number} index  - current parameter index in function signature
 * @returns {String} parameter name (generated for destructured arguments)
 */
const extractParamName = (node, index) => {
  switch (node.type) {
    case 'Identifier':
      // regular parameter
      return node.name
    case 'AssignmentPattern':
    case 'AssignmentExpression': // eslint-disable-line no-fallthrough
      // parameter with default values
      // might be a destructured parameter
      return node.left.name ? node.left.name : extractParamName(node.left, index)
    case 'RestElement':
    case 'SpreadElement': // eslint-disable-line no-fallthrough
      // rest parameter
      return node.argument.name
    case 'ObjectPattern':
    case 'ObjectExpression': // eslint-disable-line no-fallthrough
    case 'ArrayPattern': // eslint-disable-line no-fallthrough
    case 'ArrayExpression': // eslint-disable-line no-fallthrough
      // rest parameter
      return `param${index + 1}`
  }
}

/**
 * Extract declared parameter names from function signature
 *
 * Lots of different signature are supported:
 *
 * - `function` keyword, with or without function name
 * - "fat arrow" syntax, with or without parenthesis
 * - parameter with default values
 * - async functions
 *
 * Rest parameters are supported.
 * Destructured parameters are returned with a generated name
 *
 * @param {Function} fn     - analyzed function
 * @returns {Array<String>} array of declared parameters (might be empty)
 * @throws {Error} if the passed argument isn't a function, or uses unsupported syntax
 */
exports.getParamNames = fn => {
  const declaration = `(${(fn || '').toString()})`
  try {
    // parse doesn't work with function shortcut in objects, like:
    // getParamNames(({test() {}}).test)
    const { body: [node] } = acorn.parse(declaration, {
      ecmaVersion: 9
    })
    if (node.type !== 'ExpressionStatement') {
      throw new Error(`unexpected node type ${node.type}`)
    }
    const { type, params, arguments: args } = node.expression
    switch (type) {
      case 'ArrowFunctionExpression':
        // a => {
        // (a, b, c) => {
        // async (a, b, c) => {
      case 'FunctionExpression': // eslint-disable-line no-fallthrough
        // async function named(a, b, c)
        return params.map(extractParamName)
      case 'CallExpression':
        // named (a, b, c) {
        // function (a, b, c) {
        // function named(a, b, c) {
        return args.map(extractParamName)
      default:
        throw new Error(`unexpected expression ${type}`)
    }
  } catch (err) {
    throw new Error(`unsupported function ${fn}`)
  }
}

/**
 * Transform an array to an object, assigning each array item to a given property.
 *
 * Provided values that doesn't have any expected properties will be kept in the
 * resulting object, affected to a property named after the index.
 * Of course, order in both parameters is significant.
 *
 * @param {Array<Any>} array          - array mapped to an object
 * @param {Array<String>} properties  - property names used to create the object
 * @returns {Object}                  obj that contain properties specified in `properties`,
 * with corresponding values from `array`
 */
exports.arrayToObj = (array, properties) => {
  const obj = {}
  // affect first expected properties
  for (const [i, prop] of properties.entries()) {
    obj[prop] = array[i]
  }
  // includes also unexpected values
  for (let i = properties.length; i < array.length; i++) {
    obj[i] = array[i]
  }
  return obj
}

/**
 * Enrich validation error for a given API, with friendly message
 *
 * When enriching validation error for:
 * - input parameters, a 400 (Bad Request) Boom error is returned
 * - returned results, a 512 (Bad Response) Boom error is returned
 *
 * @param {Error} err               - Joi validation error, or null
 * @param {String} id                 - API id used in error detailed report
 * @param {Boolean} [forInput = true] - true when enriching error for input parameters,
 *                                    false for returned result
 * @returns {(Error|null)}            the enriched detailed error, or null err was so
 */
exports.enrichError = (err, id, forInput = true) => {
  if (!err) {
    return null
  }
  // customize error message for convenience
  err.message = `Incorrect ${forInput ? 'parameters' : 'response'} for API ${id}: ${err.message}`
  const error = boomify(err, { statusCode: forInput ? 400 : 512 })
  if (!forInput) {
    error.output.payload.error = 'Bad Response'
  }

  return error
}

/**
 * Check if the exposed apis is acceptable for further processing
 *
 * @param {*} apis      - checked apis
 * @returns {Boolean}   true if object can be exposed
 */
exports.isApi = apis => !!apis && typeof apis === 'object' && !Array.isArray(apis)

/**
 * Expected schema for exposed API groups
 * @type {Joi}
 */
exports.groupSchema = joi.object({
  name: joi.string().required(),
  init: joi.func().required()
}).unknown(true)

/**
 * Extract API groups and API options from parameter
 *
 * API groups are objects compliant with mini-service-utils.groupSchema.
 * They could be specified in the `groups` property (options will be in `groupOpts`).
 * The specify parameter could be the group itself, mixed with the options.
 *
 * @param {Object} opts       - analyzed parameter
 * @return {ExtractedGroups}  extracted groups
 */
exports.extractGroups = opts => {
  const result = { groups: [], groupOpts: {} }
  if (!exports.groupSchema.validate(opts).error) {
    // APIs are contained in the opts themselves
    result.groups.push(opts)
    result.groupOpts[opts.name] = opts
  } else if (Array.isArray(opts.groups)) {
    // APIs are grouped
    const valid = joi.array().items(exports.groupSchema).validate(opts.groups)
    if (valid.error) throw new Error(`Group definition ${valid.error.message.replace(/^"value" /, '')}`)
    result.groups = opts.groups
    result.groupOpts = opts.groupOpts || {}
  } else {
    throw new Error('No APIs nor APIs groups defined')
  }
  return result
}

/**
 * Extract validation objects for a given API
 *
 * Each exposed API can have a Joi object assigned to validate received parameters/response
 * This validation object can be either:
 * - in the clause property of the API function
 * - in the clause property of the API group options
 *
 * @param {String} id                     - name of the exposed API for which validation object is researched
 * @param {Object} apis                   - hash containing the exposed API function
 * @param {Object} opts                   - options that may contains validation hash
 * @param {Object} [opts.validate]          - May contains validation object for the exposed API
 * @param {String} [clause = 'validate']  - name of the searched clause, defaults to 'validate'
 * @return {(Joi|null)}       validation object found, or null
 */
exports.extractValidate = (id, apis, opts, clause = 'validate') =>
  apis[id][clause] || (opts && opts[clause]) || null

/**
 * HTTP header name used in response to exchange exposed API checksum.
 * Will contain the CRC-32 checksum of the `JSON.stringify(exposed.apis)` array
 * @type {String}
 */
exports.checksumHeader = 'x-service-crc'

/**
 * Expected schema for global options, with logger and transport type
 * @type {Joi}
 */
exports.optionSchema = joi.object({
  transport: joi.object({
    type: joi.string().lowercase().required().regex(/^[\w-]+$/)
  }).unknown(true).required(),
  logger: joi.object({
    debug: joi.func(),
    info: joi.func()
  }).unknown(true).requiredKeys('')
}).unknown(true)

/**
 * Validates incoming options and synchronously load transport plugin.
 * Transport plugins are expected to be in `./transports/` folder, a relative path
 * to caller's `__dirname`
 *
 * @param {Object} options    - options to validate
 * @param {Function} require  - the require function used to require transport
 * @returns {Object} loaded transport
 * @throws {Error} when transport cannot be requiredt found
 */
exports.loadTransport = (options, require) => {
  joi.assert(options, exports.optionSchema)
  const { logger, transport: { type } } = options
  try {
    logger.debug(`Load transport ${type}`)
    return require(`./transports/${type}`)
  } catch (err) {
    throw new Error(`Cannot load transport ${type}: ${err.message}`)
  }
}
